### ISSUE TITLE
Default Arguments in Atom Constructors

### ISSUE THREAD

**Issue by 4e6** [enso-org/enso#1600]
### General Summary

Consider the following program:

```py
from Builtins import all

type A x=42

g x=42 = x

main =
    f x=42 = x

    IO.println A
    IO.println f
    IO.println here.g
```

Output:
```
Constructor<A>
org.enso.interpreter.runtime.callable.function.Function@255ed8e4
42
```

Both the atom constructor `A` and the function `f` don't apply their default arguments and return functions.

### Steps to Reproduce

Run the example.

### Expected Result

Atom constructor `A` and function `f` should apply their default arguments and return atom `A 42` and `42` respectively.

### Actual Result

Atom constructor `A` and function `f` return functions.

### Enso Version

main @ efe0e859e7a8af6265ccc1b3998ad0e03d1bb828

**Comment by iamrecursion** [comment]
I think you just might need to add `.call` to the values. We don't currently have any mechanism to say whether a function value should be forced rather than left suspended if it's pre-applied with defaults. Improvements to this are tracked in #473.

**Comment by 4e6** [comment]
I checked. You can `call` the function, but not the atom

```py
from Builtins import all

type A x=42

g x=42 = x

main =
    f x=42 = x

    IO.println f.call
    IO.println A.call
```

```
42
Execution finished with an error: Method `call` of A could not be found.
        at <enso> argument<1>(src/Main.enso:11:16-21)
        at <enso> Main.main(src/Main.enso:11:5-21)
```

**Comment by iamrecursion** [comment]
Okay. That's the real bug here. It's probably that we don't register the equivalent on the atom. It should be simple to fix if you feel up to it.

**Comment by radeusgd** [comment]
If I understand correctly, this issue is also related to default arguments not working within Atoms, right?

As in, we have atoms like
```
type Matching_Strategy
    type Exact case_sensitivity=True
```

and for `foo : Matching_Strategy -> Text -> Text` we'd like to be able to write `foo Exact "bar"`. But right now this will not work, `Exact` will be treated as an AtomConstructor and not an Atom, so we'll get a type error. We need to write `foo (Exact True) "bar"` as if the default argument was not there.

**Comment by radeusgd** [comment]
I hope what I'm writing is on topic here, as I understand that this issue is the wider 'Atoms with default arguments currently have problems'. If not - please correct me and I'll create a separate issue.

Working with atoms I've found out that if the default arguments are primitives, then they work except for the first argument (i.e. if you provide 0 args you get an AtomConstructor, but if you provide at least one, the others get defaulted correctly).

But for some reason if the default argument is itself a composite Atom, the interpreter crashes :confused: 

Best to illustrate it on this small example:
```
from Standard.Base import all

type Case_Insensitive
type Matching_Strategy
    type Exact (cs : (True | Case_Insensitive) = True)

type Column_Selector
    type By_Name (names : Vector Text) (ms : Matching_Strategy = Exact True)

# Interestingly, if there are multiple arguments but the default is a primitive, it seems to work:
type Multi bar=0 baz=42

main =
    y = Multi 23
    IO.println y

    z = Multi
    IO.println z

    w = Multi ["bar"]
    IO.println w
    
    # But this will break hard
    x = By_Name ["Foo"] # note the last argument is missing
    IO.println x
```

which produces the following output:

```
(Multi 23 42)
Constructor<Multi>
(Multi ['bar'] 42)
Execution finished with an error: java.lang.ClassCastException: class org.enso.interpreter.node.expression.builtin.InstantiateAtomNode cannot be cast to class org.enso.interpreter.node.EnsoRootNode (org.enso.interpreter.node.expression.builtin.InstantiateAtomNode and org.enso.interpreter.node.EnsoRootNode are in unnamed module of loader com.oracle.graalvm.locator.GraalVMLocator$GuestLangToolsLoader @7ef27d7f)
        at <java> org.enso.interpreter.node.BaseNode.getStateFrameSlot(BaseNode.java:35)
        at <java> org.enso.interpreter.node.callable.ApplicationNode.executeGeneric(ApplicationNode.java:97)
        at <java> org.enso.interpreter.node.callable.argument.ReadArgumentNode.executeGeneric(ReadArgumentNode.java:56)
        at <java> org.enso.interpreter.node.expression.atom.InstantiateNode.executeGeneric(InstantiateNode.java:60)
        at <java> org.enso.interpreter.node.expression.builtin.InstantiateAtomNode.execute(InstantiateAtomNode.java:33)
        at <java> org.enso.interpreter.node.expression.builtin.InstantiateAtomNode.execute(InstantiateAtomNode.java:12)
        at <enso> Main.main(src/Main.enso:24:9-23)
```

**Comment by radeusgd** [comment]
Renamed the issue to reflect more the main issue we have here - we want default arguments in Atom constructors to behave in the same way as they do for regular functions.

**Comment by radeusgd** [comment]
I wanted to clarify one possible uncertainty related to the design of default arguments in constructors.

```
type Arg x=1 y="a"

z = Arg
foo z
```

When passing the value `z` to `foo` - how do we know if `Arg` should be treated as an `AtomConstructor` (we sometimes do pass constructors directly, for example in `Test` `should_fail_with Constructor`) or should it be a constructed Atom with the default arguments applied?

Technically sometimes we could infer that from the type, but I don't think we can do it right now and even more so, sometimes the type can be so broad (`Any`) that still both variants would match it. What would be the course of action in this case?

For context, I think there is a similar thing with functions:
```
fun x=4 y="b" = ...

z = fun
foo z
```

But there we can easily disambiguate - by default `z` can be defaulted to applying the default arguments and if we want to pass the unapplied function we can instead fo `z = fun _ _`.

But this does not work with Atoms, I think - because `Arg _ _` will become a two-argument lambda constructing the atom `Arg` - effectively equivalent to the `Arg` constructor, but it is not the `Arg` constructor and we cannot for example do `Arg 1 2 . is_a (Arg _ _)` while, I think (?), we want to be able to do `Arg 1 2 . is_a Arg`.

**Comment by kustosz** [comment]
Here come some thoughts before we start the implementation:
1. For Atoms we should be able to discover the need for applying defaults statically – we already know what atom is being used, so the compiler should see that it's not in an application position but should be.
2. The deeper issue for local functions (and, more importantly, functions received as arguments). There seems to be no way to do this other than do a runtime check every time we read from a frame. Not the fastest way to do things, but it fixes a semantic bug and therefore takes precedence over performance. With good profiling, we should be able to not take a huge hit in compiled code too. This approach was discussed with @wdanilo 
3. We need to update the `...` operator, such that it identically returns its arguments (if no other arguments passed), such that if `Foo` is a constructor, `Foo ...` is still a constructor. Currently this would unwrap to the constructor function. This will be important for things like `Meta.is_a (Foo ...) (Foo 20)` if `Foo` has its arguments defaulted.
4. The CCE mentioned above needs to be tracked down and addressed.
